ct/l1: metastore frontend boilerplate

andrwng opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
There's ongoing work to implement a metastore backed by an internal topic. The leaders of the topic partitions will serve metadata backed by a persisted STM. To this end, this PR introduces RPCs and a frontend that will route to the topic partitions' leaders.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng committed:
ct/l1: add RPC request/reply types

Adds request/reply types that will be used for an upcoming frontend that
will route to partition leaders of the metastore topic.

The types here reflect the metastore interface, as the upcoming frontend
will have RPCs that map 1:1 with metastore interface methods.

andrwng committed:
ct/l1: add metastore frontend boilerplate

This adds the boilerplate required for the metastore frontend, that
routes RPCs to the metastore partition leader's shard. This consists
generally of three methods per underlying metastore call:
- do_FOO: performs the actual FOO operation on the local shard.
- FOO_locally: routes the request to the appropriate shard for the given
  FOO request and calls do_FOO on that shard.
- FOO: determines whether the target partition for the request has a
  leader that is local. If so, calls FOO_locally. Otherwise dispatches
  an FOO RPC to the leader.

This pattern, along with the concepts and templatized code for handling
routing, is copied almost verbatim from datalake/coordinator/frontend.*,
with the caveat that the concepts to verify request types are changed to
use topic_id_partition instead of topic name.

do_FOO methods are implemented as stubs, to be filled out in subsequent
commits.

andrwng committed:
model: add L1 metastore topic constants

dotnwat commented on src/v/model/namespace.h:
```diff
@@ -86,6 +86,10 @@ inline const model::topic datalake_coordinator_topic("datalake_coordinator");
 
+inline const model::topic l1_metastore_topic("ct_l1_domain");
+inline const model::topic_namespace
+  l1_metastore_nt(model::kafka_internal_namespace, model::l1_metastore_topic);
```
should it go into the "redpanda" namespace?

andrwng replied:
There was some instability noted with the transforms coordinator when using "redpanda", my guess is due to how hacky the storage layer is/was w.r.t enabling/disabling housekeeping based on special namespaces... Maybe that's gone now, but I'd rather stick to the status quo

rockwotj replied:
Agreed. I think where we settled is `redpanda` namespace is only for really special NTPs like the controller or the kvstore and all other internal stuff goes into `kafka_internal`. I mean personally I think we really only need a single internal namespace, but here we are.

dotnwat replied:
sgtm. missed this conversation wherever it happened

andrwng replied:
This thread has a little more color, discussed in the context of Iceberg https://github.com/redpanda-data/redpanda/pull/23471#discussion_r1777445276

dotnwat commented on src/v/cloud_topics/level_one/metastore/rpc.json:
```diff
@@ -0,0 +1,39 @@
+            "input_type": "get_offsets_request",
+            "output_type": "get_offsets_reply"
+        },
+        {
+            "name": "get_compaction_offsets",
+            "input_type": "get_compaction_offsets_request",
+            "output_type": "get_compaction_offsets_reply"
+        }
```
🔥

dotnwat commented on src/v/cloud_topics/level_one/metastore/frontend.cc:
```diff
@@ -0,0 +1,409 @@
+    write(temp, tp);
+    auto bytes = iobuf_to_bytes(temp);
+    auto partition = murmur2(bytes.data(), bytes.size())
+                     % md->get().get_configuration().partition_count;
```
we'll want to prevent partition count from being alterned on this topic. we might also want to do something defensive where we "register" new topic/partitions in a metastore partition and then verify that we aren't getting any misdirected RPCs.

andrwng replied:
Yeah, I'm thinking we might need the domain managers to listen for partition count changes, and then respond by replicating to the STM some record of that change to begin migrating metadata or something... Generally seems like we will definitely need some way to deterministically tell if parts of the STM are being migrated as a result of new partitions

rockwotj replied:
I wonder if we should just explicitly hardcode to partition 0 for now 🤔

dotnwat replied:
I don't think we should handle changing number of partitions right now. Ideally we can tackle that when we tackle the broader load balancing challege. But you're right andrew, if we really needed to do it, we could track the history of the partition count changes. then the partition function would determine which partition count was used when a partition was created (or something along these lines).

andrwng replied:
I don't think there's harm in having multiple partitions right now -- it's a hopefully painless, assuming no migrations, thing to keep in, while giving us a real-ish idea of how domains will eventually work. I can change it if you feel strongly about it though.

Migrations OTOH, I think we should punt on, until maybe even after the first cut

rockwotj replied:
I don't feel strongly about anything other than punting on migrations :) I'm fine with allowing multiple partitions to start, but just needs clear expectations it can't be changed.

dotnwat commented on src/v/cloud_topics/level_one/metastore/frontend.cc:
```diff
@@ -0,0 +1,409 @@
+}
+
+std::optional<model::partition_id>
+frontend::metastore_partition(const model::topic_id_partition& tp) const {
```
is this called from somewhere?

this part above was confusing to me

```
        if constexpr (request_has_topic_id_partition<req_t>) {
            // XXX: hash
            metastore_pid = req.tp.partition;
        } else {
            metastore_pid = req.metastore_partition;
        }
```

andrwng replied:
Ah oops! Fixed

dotnwat reviewed (dismissed):
no real complaints here

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--a93b174e-c2e2-4d1a-aacc-2f10094f78fd-->
<!--TEST RESULTS-->
<!--4de0a1f8-096e-4d0f-b54f-28b980f6c0d3-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69994">build#69994</a></summary>


| test_class       | test_method              | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------|:-------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeE2ETests | test_json_schema_unicode | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"} | integration | https://buildkite.com/redpanda/redpanda/builds/69994#01985ea4-2909-45fb-a913-a75250c8c821 | FLAKY         | 20/21    | upstream reliability is '99.10514541387025'. current run reliability is '95.23809523809523'. drift is 3.86705 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70027">build#70027</a></summary>


| test_class               | test_method              | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:-------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeE2ETests         | test_json_schema_unicode | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"}                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/70027#019861c1-ba00-42d8-ad4d-ef89fe587332 | FLAKY         | 19/21    | upstream reliability is '99.08675799086758'. current run reliability is '90.47619047619048'. drift is 8.61057 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/70027#019861c1-b9fd-4f18-afc5-bb224b51b6ae | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<!--/body-->

rockwotj commented on src/v/cloud_topics/level_one/metastore/frontend.cc:
```diff
@@ -0,0 +1,413 @@
+
+ss::future<rpc::get_first_offset_ge_reply> do_get_first_offset_ge(
+  cluster::partition_manager&, model::ntp, rpc::get_first_offset_ge_request) {
+    co_return rpc::get_first_offset_ge_reply{.ec = rpc::errc::timed_out};
```
why are the above OK and this timed_out?

andrwng replied:
Heh no good reason. I initially tried/failed to get claude to generate some of this code. This is an artifact of its attempt + me copy pasting what it wrote.

rockwotj commented on src/v/cloud_topics/level_one/metastore/frontend.cc:
```diff
@@ -0,0 +1,413 @@
+      default_metastore_partitions,
+      replication_factor};
+
+    topic.properties.compression = model::compression::none;
```
this topic property does nothing

andrwng replied:
Done

rockwotj commented on src/v/cloud_topics/level_one/metastore/frontend.cc:
```diff
@@ -0,0 +1,413 @@
+}
+auto frontend::process(req_t req, bool local_only) {
+    using resp_t = req_t::resp_t;
+    return ensure_topic_exists().then([req = std::move(req), local_only, this](
```
No coro love? :)

But we really should make this process simpler... 🤔

andrwng replied:
Done

rockwotj reviewed (dismissed):
LGTM, just some nits and thoughts to consider

andrwng committed:
ct/l1: coroutinize metastore frontend templates

Coroutinizes the templates in the metastore frontend. The code was
written with `auto` originally, which doesn't appear to love being mixed
with coroutines, so this also explicitly specifies the reply types.

andrwng committed:
ct/l1: wrap metastore frontend in RPC service

This is the RPC service that eventually needs to be instantiated by the
Redpanda application and managed by the RPC server.

andrwng commented:
Sorry, one more push for one new commit of boilerplate code
